### PR TITLE
Fieldmap

### PR DESCRIPTION
**Pull request by kjvbrt**
This PR implements magnetic field described with the help of the fieldmap. Requires 6 input vectors: position: x, y, z and filed: Bx, By, Bz.

### CODE REVIEW

File: SimG4Components/src/SimG4MagneticFieldFromMapTool.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  Gaudi::Property<double> m_maxEps{this, "MaximumEpsilon", 0, "Maximum epsilon (see G4 documentation)"};
+  /// This parameter governs accuracy of volume intersection, see G4 doc for more details. Set with property DeltaChord
+  Gaudi::Property<double> m_deltaChord{this, "DeltaChord", 0, "Missing distance for the chord finder"};
+  /// This parameter is roughly the position error which is acceptable in an integration step, see G4 doc for details.
+  /// Set with property DeltaOneStep
+  Gaudi::Property<double> m_deltaOneStep{this, "DeltaOneStep", 0, "Delta(one-step)"};
+  /// Upper limit of the step size, see G4 doc for more details. Set with property MaximumStep
+  Gaudi::Property<double> m_maxStep{this, "MaximumStep", 1. * m, "Maximum step length in field (see G4 documentation)"};
+  /// Lower limit of the step size, see G4 doc for more details. Set with property MaximumStep
+  Gaudi::Property<double> m_minStep{this, "MinimumStep", 0.01 * mm, "Maximum step length in field (see G4 documentation)"};
```
Comment (BrieucF) [inline SimG4Components/src/SimG4MagneticFieldFromMapTool.h:75]: Maximum --> Minimum

File: SimG4Components/src/SimG4MagneticFieldFromMapTool.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  /// Name of the integration stepper, defaults to NystromRK4.
+  Gaudi::Property<std::string> m_integratorStepper{this, "IntegratorStepper", "NystromRK4", "Integrator stepper name"};
+  /// Path to the input file containing fieldmap
+  Gaudi::Property<std::string> m_mapFilePath{this, "MapFile", "", "Path to file containing fieldmap"};
+  /// Additional constant field, z component (spans whole z range of the map)
+  Gaudi::Property<double> m_addFieldBz{this, "AddFieldBz", 0., "Additional constant field, z component"};
+  /// Maximum radius of the additional constant field (default: no limit)
+  Gaudi::Property<double> m_addFieldMaxR{this, "AddFieldMaxR", -1., "Maximum radius of additional constant field (default: no limit)"};
```
Comment (BrieucF) [inline SimG4Components/src/SimG4MagneticFieldFromMapTool.h:83]: Shall we also have a property controlling the maximal radius at which we extract field values from the map (maps from COMSOL may have a large unneeded radial extent) or is it negligible in resource consumption? Maybe the same variable could be used for the constant field and the field from the map?

File: SimG4Components/src/SimG4MagneticFieldFromMapTool.cpp [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+    if (!sc.isSuccess()) {
+      return sc;
+    }
+  } else if (m_mapFilePath.value().find(".txt") != std::string::npos) {
+    sc = loadComsolMap();
+    if (!sc.isSuccess()) {
+      return sc;
+    }
+  } else {
+    error() << "Fieldmap file extension not recognized!" << endmsg;
```
Comment (BrieucF) [inline SimG4Components/src/SimG4MagneticFieldFromMapTool.cpp:78]: Maybe you could explicitly print here that the algo expects a .root or .txt

**Review by BrieucF — COMMENTED** [review]
Very nice, thanks! Only very minor things

File: SimG4Components/src/SimG4MagneticFieldFromMapTool.h (hunk 1 above)
Comment (kjvbrt) [reply SimG4Components/src/SimG4MagneticFieldFromMapTool.h]: Fixed

File: SimG4Components/src/SimG4MagneticFieldFromMapTool.h (hunk 2 above)
Comment (kjvbrt) [reply SimG4Components/src/SimG4MagneticFieldFromMapTool.h:83]: Added maximal R and z of the map for both 2d and 3d map fields.

File: SimG4Components/src/SimG4MagneticFieldFromMapTool.cpp (hunk 3 above)
Comment (kjvbrt) [reply SimG4Components/src/SimG4MagneticFieldFromMapTool.cpp:78]: Add list of recognized extensions.

**Comment by BrieucF** [comment]
Great thanks! +1 on my side

**Pull request merged**